Compiler crashes in ExpandSAMs phase

## Compiler version

```
$ sbt "print scalaVersion"
[snip]
scala3-library / scalaVersion
        3.2.0-RC3
scala3-interfaces / scalaVersion
        3.2.0-RC3
tasty-core / scalaVersion
        3.2.0-RC3
scala3-sbt-bridge / scalaVersion
        3.2.0-RC3
scala3-compiler / scalaVersion
        3.2.0-RC3
scalaVersion
        3.2.0-RC3
```

```
$ java -version
openjdk version "11.0.16" 2022-07-19
OpenJDK Runtime Environment (build 11.0.16+8-post-Debian-1deb11u1)
OpenJDK 64-Bit Server VM (build 11.0.16+8-post-Debian-1deb11u1, mixed mode, sharing)
```

```
$ git log -n 1 | cat
commit e560c2d6970c864ceea5607597d74cf49c443965
Merge: 44a2f7292f 40c4fe3d8c
Author: Paweł Marks <pmarks@virtuslab.com>
Date:   Fri Aug 5 07:35:30 2022 -0700

    Merge pull request #15825 from tgodzik/dont-drop-type-params

    bugfix: Retain partial type params typying information on error
```

## Minimized code

```scala
// crash.scala
import scala.language.implicitConversions

class MyFunction(args: String*)

trait MyFunction0[+R] extends MyFunction {
  def apply(): R
}

implicit def fromFunction0[R](f: Function0[R]): MyFunction0[R] = () => f()
```

## Output

```
$ sbt "scalac crash.scala"
[snip]
exception occurred while compiling crash.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: asTerm called on not-a-Term val <none>
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
        at dotty.tools.dotc.core.Symbols$Symbol.asTerm(Symbols.scala:169)
        at dotty.tools.dotc.ast.tpd$.ClassDef(tpd.scala:309)
        at dotty.tools.dotc.ast.tpd$.AnonClass(tpd.scala:370)
        at dotty.tools.dotc.ast.tpd$.AnonClass(tpd.scala:352)
        at dotty.tools.dotc.transform.ExpandSAMs.transformBlock(ExpandSAMs.scala:63)
        at dotty.tools.dotc.transform.MegaPhase.goBlock(MegaPhase.scala:740)
        at dotty.tools.dotc.transform.MegaPhase.goBlock(MegaPhase.scala:741)
        at dotty.tools.dotc.transform.MegaPhase.transformBlock(MegaPhase.scala:450)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:298)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.mapDefDef$1(MegaPhase.scala:248)
        at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:251)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:425)
        at dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:438)
        at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:438)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:359)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.transformNamed$1(MegaPhase.scala:255)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:425)
        at dotty.tools.dotc.transform.MegaPhase.loop$1(MegaPhase.scala:438)
        at dotty.tools.dotc.transform.MegaPhase.transformStats(MegaPhase.scala:438)
        at dotty.tools.dotc.transform.MegaPhase.mapPackage$1(MegaPhase.scala:379)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:382)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:427)
        at dotty.tools.dotc.transform.MegaPhase.transformUnit(MegaPhase.scala:454)
        at dotty.tools.dotc.transform.MegaPhase.run(MegaPhase.scala:466)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:311)
        at scala.collection.immutable.List.map(List.scala:246)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:312)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:234)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:245)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:253)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:262)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:262)
        at dotty.tools.dotc.Run.compileSources(Run.scala:186)
        at dotty.tools.dotc.Run.compile(Run.scala:170)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
        at dotty.tools.dotc.Driver.process(Driver.scala:195)
        at dotty.tools.dotc.Driver.process(Driver.scala:163)
        at dotty.tools.dotc.Driver.process(Driver.scala:175)
        at dotty.tools.dotc.Driver.main(Driver.scala:205)
        at dotty.tools.dotc.Main.main(Main.scala)
java.lang.AssertionError: assertion failed: asTerm called on not-a-Term val <none> while compiling crash.scala
[error] Nonzero exit code returned from runner: 1
[error] (scala3-compiler / Compile / runMain) Nonzero exit code returned from runner: 1
[error] Total time: 2 s, completed 2022/08/14 18:36:16
```

## Expectation

dotc doesn't crash.

## Notes

* Scala 3.1.3 is also facing this issue
* I have extracted the minimized code from [scala-js's library code](https://github.com/scala-js/scala-js/blob/112def5cb59b942d8b2c4c5791b17ea0c7c4a7dc/library/src/main/scala-new-collections/scala/scalajs/js/Any.scala#L135)
  * So, you will face the same issue when you try to compile scala-js library with dotty

I believe I got another instance of this issue:

```sh
➜  ~/scala-sam-bug cat test2.scala
```

```scala
//> using dep org.scala-js::scalajs-dom_sjs1:2.8.0
import org.scalajs.dom.document
@main def test2 = document.getElementById("foo").addEventListener("input", _ => ???)
``` 

```sh
➜  ~/scala-sam-bug scala-cli compile test2.scala
Compiling project (Scala 3.3.1, JVM (17))
Error compiling project (Scala 3.3.1, JVM (17))
Error: Unexpected error when compiling scala-sam-bug_035d174876-089f72375e: java.lang.AssertionError: assertion failed: asTerm called on not-a-Term val <none>
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
        at dotty.tools.dotc.core.Symbols$Symbol.asTerm(Symbols.scala:169)
        at dotty.tools.dotc.ast.tpd$.ClassDef(tpd.scala:324)
        at dotty.tools.dotc.ast.tpd$.AnonClass(tpd.scala:385)
        at dotty.tools.dotc.ast.tpd$.AnonClass(tpd.scala:367)
        at dotty.tools.dotc.transform.ExpandSAMs.transformBlock(ExpandSAMs.scala:63)
        at dotty.tools.dotc.transform.MegaPhase.goBlock(MegaPhase.scala:759)
        at dotty.tools.dotc.transform.MegaPhase.goBlock(MegaPhase.scala:760)
        at dotty.tools.dotc.transform.MegaPhase.transformBlock(MegaPhase.scala:465)
        at dotty.tools.dotc.transform.MegaPhase.transformUnnamed$1(MegaPhase.scala:303)
        at dotty.tools.dotc.transform.MegaPhase.transformTree(MegaPhase.scala:442)
        at dotty.tools.dotc.transform.MegaPhase.loop$3(MegaPhase.scala:474)
        ...
```

--- a/compiler/src/dotty/tools/dotc/ast/tpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/tpd.scala
@@ -328,21 +328,6 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
       superArgs: List[Tree] = Nil, adaptVarargs: Boolean = false)(using Context): TypeDef =
     val firstParent :: otherParents = cls.info.parents: @unchecked
 
-    def isApplicable(constr: Symbol): Boolean =
-      def recur(ctpe: Type): Boolean = ctpe match
-        case ctpe: PolyType =>
-          recur(ctpe.instantiate(firstParent.argTypes))
-        case ctpe: MethodType =>
-          var paramInfos = ctpe.paramInfos
-          if adaptVarargs && paramInfos.length == superArgs.length + 1
-            && atPhaseNoLater(Phases.elimRepeatedPhase)(constr.info.isVarArgsMethod)
-          then // accept missing argument for varargs parameter
-            paramInfos = paramInfos.init
-          superArgs.corresponds(paramInfos)(_.tpe <:< _)
-        case _ =>
-          false
-      recur(constr.info)
-
     def adaptedSuperArgs(ctpe: Type): List[Tree] = ctpe match
       case ctpe: PolyType =>
         adaptedSuperArgs(ctpe.instantiate(firstParent.argTypes))
@@ -357,8 +342,11 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
     val superRef =
       if cls.is(Trait) then TypeTree(firstParent)
       else
-        val constr = firstParent.decl(nme.CONSTRUCTOR).suchThat(isApplicable)
-        New(firstParent, constr.symbol.asTerm, adaptedSuperArgs(constr.info))
+        val parentConstr = firstParent.applicableConstructors(superArgs.tpes, adaptVarargs) match
+          case Nil => assert(false, i"no applicable parent constructor of $firstParent for supercall arguments $superArgs")
+          case constr :: Nil => constr
+          case _ => assert(false, i"multiple applicable parent constructors of $firstParent for supercall arguments $superArgs")
+        New(firstParent, parentConstr.asTerm, adaptedSuperArgs(parentConstr.info))
 
     ClassDefWithParents(cls, constr, superRef :: otherParents.map(TypeTree(_)), body)
   end ClassDef

--- a/compiler/src/dotty/tools/dotc/ast/tpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/tpd.scala
@@ -315,24 +315,53 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
   def TypeDef(sym: TypeSymbol)(using Context): TypeDef =
     ta.assignType(untpd.TypeDef(sym.name, TypeTree(sym.info)), sym)
 
-  def ClassDef(cls: ClassSymbol, constr: DefDef, body: List[Tree], superArgs: List[Tree] = Nil)(using Context): TypeDef = {
+  /** Create a class definition
+   *  @param cls          the class symbol of the created class
+   *  @param constr       its primary constructor
+   *  @param body         the statements in its template
+   *  @param superArgs    the arguments to pass to the superclass constructor
+   *  @param adaptVarargs if true, allow matching a vararg superclass constructor
+   *                      with a missing argument in superArgs, and synthesize an
+   *                      empty repeated parameter in the supercall in this case
+   */
+  def ClassDef(cls: ClassSymbol, constr: DefDef, body: List[Tree],
+      superArgs: List[Tree] = Nil, adaptVarargs: Boolean = false)(using Context): TypeDef =
     val firstParent :: otherParents = cls.info.parents: @unchecked
+
+    def isApplicable(constr: Symbol): Boolean =
+      def recur(ctpe: Type): Boolean = ctpe match
+        case ctpe: PolyType =>
+          recur(ctpe.instantiate(firstParent.argTypes))
+        case ctpe: MethodType =>
+          var paramInfos = ctpe.paramInfos
+          if adaptVarargs && paramInfos.length == superArgs.length + 1
+            && atPhaseNoLater(Phases.elimRepeatedPhase)(constr.info.isVarArgsMethod)
+          then // accept missing argument for varargs parameter
+            paramInfos = paramInfos.init
+          superArgs.corresponds(paramInfos)(_.tpe <:< _)
+        case _ =>
+          false
+      recur(constr.info)
+
+    def adaptedSuperArgs(ctpe: Type): List[Tree] = ctpe match
+      case ctpe: PolyType =>
+        adaptedSuperArgs(ctpe.instantiate(firstParent.argTypes))
+      case ctpe: MethodType
+      if ctpe.paramInfos.length == superArgs.length + 1 =>
+        // last argument must be a vararg, otherwise isApplicable would have failed
+        superArgs :+
+          repeated(Nil, TypeTree(ctpe.paramInfos.last.argInfos.head, inferred = true))
+      case _ =>
+        superArgs
+
     val superRef =
-      if (cls.is(Trait)) TypeTree(firstParent)
-      else {
-        def isApplicable(ctpe: Type): Boolean = ctpe match {
-          case ctpe: PolyType =>
-            isApplicable(ctpe.instantiate(firstParent.argTypes))
-          case ctpe: MethodType =>
-            (superArgs corresponds ctpe.paramInfos)(_.tpe <:< _)
-          case _ =>
-            false
-        }
-        val constr = firstParent.decl(nme.CONSTRUCTOR).suchThat(constr => isApplicable(constr.info))
-        New(firstParent, constr.symbol.asTerm, superArgs)
-      }
+      if cls.is(Trait) then TypeTree(firstParent)
+      else
+        val constr = firstParent.decl(nme.CONSTRUCTOR).suchThat(isApplicable)
+        New(firstParent, constr.symbol.asTerm, adaptedSuperArgs(constr.info))
+
     ClassDefWithParents(cls, constr, superRef :: otherParents.map(TypeTree(_)), body)
-  }
+  end ClassDef
 
   def ClassDefWithParents(cls: ClassSymbol, constr: DefDef, parents: List[Tree], body: List[Tree])(using Context): TypeDef = {
     val selfType =
@@ -359,13 +388,18 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
    *  @param parents        a non-empty list of class types
    *  @param termForwarders a non-empty list of forwarding definitions specified by their name and the definition they forward to.
    *  @param typeMembers    a possibly-empty list of type members specified by their name and their right hand side.
+   *  @param adaptVarargs   if true, allow matching a vararg superclass constructor
+   *                        with a missing argument in superArgs, and synthesize an
+   *                        empty repeated parameter in the supercall in this case
    *
    *  The class has the same owner as the first function in `termForwarders`.
    *  Its position is the union of all symbols in `termForwarders`.
    */
-  def AnonClass(parents: List[Type], termForwarders: List[(TermName, TermSymbol)],
-      typeMembers: List[(TypeName, TypeBounds)] = Nil)(using Context): Block = {
-    AnonClass(termForwarders.head._2.owner, parents, termForwarders.map(_._2.span).reduceLeft(_ union _)) { cls =>
+  def AnonClass(parents: List[Type],
+      termForwarders: List[(TermName, TermSymbol)],
+      typeMembers: List[(TypeName, TypeBounds)],
+      adaptVarargs: Boolean)(using Context): Block = {
+    AnonClass(termForwarders.head._2.owner, parents, termForwarders.map(_._2.span).reduceLeft(_ union _), adaptVarargs) { cls =>
       def forwarder(name: TermName, fn: TermSymbol) = {
         val fwdMeth = fn.copy(cls, name, Synthetic | Method | Final).entered.asTerm
         for overridden <- fwdMeth.allOverriddenSymbols do
@@ -385,6 +419,9 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
    * with the specified owner and position.
    */
   def AnonClass(owner: Symbol, parents: List[Type], coord: Coord)(body: ClassSymbol => List[Tree])(using Context): Block =
+    AnonClass(owner, parents, coord, adaptVarargs = false)(body)
+
+  private def AnonClass(owner: Symbol, parents: List[Type], coord: Coord, adaptVarargs: Boolean)(body: ClassSymbol => List[Tree])(using Context): Block =
     val parents1 =
       if (parents.head.classSymbol.is(Trait)) {
         val head = parents.head.parents.head
@@ -393,7 +430,7 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
       else parents
     val cls = newNormalizedClassSymbol(owner, tpnme.ANON_CLASS, Synthetic | Final, parents1, coord = coord)
     val constr = newConstructor(cls, Synthetic, Nil, Nil).entered
-    val cdef = ClassDef(cls, DefDef(constr), body(cls))
+    val cdef = ClassDef(cls, DefDef(constr), body(cls), Nil, adaptVarargs)
     Block(cdef :: Nil, New(cls.typeRef, Nil))
 
   def Import(expr: Tree, selectors: List[untpd.ImportSelector])(using Context): Import =

--- a/compiler/src/dotty/tools/dotc/core/Phases.scala
+++ b/compiler/src/dotty/tools/dotc/core/Phases.scala
@@ -532,7 +532,7 @@ object Phases {
   def sbtExtractAPIPhase(using Context): Phase          = ctx.base.sbtExtractAPIPhase
   def picklerPhase(using Context): Phase                = ctx.base.picklerPhase
   def inliningPhase(using Context): Phase               = ctx.base.inliningPhase
-  def stagingPhase(using Context): Phase               = ctx.base.stagingPhase
+  def stagingPhase(using Context): Phase                = ctx.base.stagingPhase
   def splicingPhase(using Context): Phase               = ctx.base.splicingPhase
   def firstTransformPhase(using Context): Phase         = ctx.base.firstTransformPhase
   def refchecksPhase(using Context): Phase              = ctx.base.refchecksPhase

--- a/compiler/src/dotty/tools/dotc/core/TypeUtils.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeUtils.scala
@@ -191,8 +191,8 @@ class TypeUtils:
       case self: RefinedOrRecType => self.parent.stripRefinement
       case seld => self
 
-    /** The constructors of this tyoe that that are applicable to `argTypes`, without needing
-     *  an implicit conversion.
+    /** The constructors of this type that are applicable to `argTypes`, without needing
+     *  an implicit conversion. Curried constructors are always excluded.
      *  @param adaptVarargs   if true, allow a constructor with just a varargs argument to
      *                        match an empty argument list.
      */
@@ -208,7 +208,7 @@ class TypeUtils:
               && atPhaseNoLater(Phases.elimRepeatedPhase)(constr.info.isVarArgsMethod)
             then // accept missing argument for varargs parameter
               paramInfos = paramInfos.init
-            argTypes.corresponds(paramInfos)(_ <:< _)
+            argTypes.corresponds(paramInfos)(_ <:< _) && !ctpe.resultType.isInstanceOf[MethodType]
           case _ =>
             false
         recur(constr.info)

--- a/compiler/src/dotty/tools/dotc/core/TypeUtils.scala
+++ b/compiler/src/dotty/tools/dotc/core/TypeUtils.scala
@@ -8,6 +8,7 @@ import Names.{Name, TermName}
 import Constants.Constant
 
 import Names.Name
+import StdNames.nme
 import config.Feature
 
 class TypeUtils:
@@ -189,5 +190,30 @@ class TypeUtils:
     def stripRefinement: Type = self match
       case self: RefinedOrRecType => self.parent.stripRefinement
       case seld => self
+
+    /** The constructors of this tyoe that that are applicable to `argTypes`, without needing
+     *  an implicit conversion.
+     *  @param adaptVarargs   if true, allow a constructor with just a varargs argument to
+     *                        match an empty argument list.
+     */
+    def applicableConstructors(argTypes: List[Type], adaptVarargs: Boolean)(using Context): List[Symbol] =
+      def isApplicable(constr: Symbol): Boolean =
+        def recur(ctpe: Type): Boolean = ctpe match
+          case ctpe: PolyType =>
+            if argTypes.isEmpty then recur(ctpe.resultType) // no need to know instances
+            else recur(ctpe.instantiate(self.argTypes))
+          case ctpe: MethodType =>
+            var paramInfos = ctpe.paramInfos
+            if adaptVarargs && paramInfos.length == argTypes.length + 1
+              && atPhaseNoLater(Phases.elimRepeatedPhase)(constr.info.isVarArgsMethod)
+            then // accept missing argument for varargs parameter
+              paramInfos = paramInfos.init
+            argTypes.corresponds(paramInfos)(_ <:< _)
+          case _ =>
+            false
+        recur(constr.info)
+
+      self.decl(nme.CONSTRUCTOR).altsWith(isApplicable).map(_.symbol)
+
 end TypeUtils
 

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -5945,16 +5945,21 @@ object Types extends TypeUtils {
     def samClass(tp: Type)(using Context): Symbol = tp match
       case tp: ClassInfo =>
         def zeroParams(tp: Type): Boolean = tp.stripPoly match
-          case mt: MethodType => mt.paramInfos.isEmpty && !mt.resultType.isInstanceOf[MethodType]
+          case mt: MethodType =>
+            val noArgsNeeded = mt.paramInfos match
+              case Nil => true
+              case info :: Nil => info.isRepeatedParam
+              case _ => false
+            noArgsNeeded && !mt.resultType.isInstanceOf[MethodType]
           case et: ExprType => true
           case _ => false
-        val cls = tp.cls
-        val validCtor =
+        def validCtor(cls: Symbol): Boolean =
           val ctor = cls.primaryConstructor
-          // `ContextFunctionN` does not have constructors
-          !ctor.exists || zeroParams(ctor.info)
-        val isInstantiable = !cls.isOneOf(FinalOrSealed) && (tp.appliedRef <:< tp.selfType)
-        if validCtor && isInstantiable then tp.cls
+          (!ctor.exists || zeroParams(ctor.info)) // `ContextFunctionN` does not have constructors
+          && (!cls.is(Trait) || validCtor(cls.info.parents.head.classSymbol))
+        def isInstantiable =
+          !tp.cls.isOneOf(FinalOrSealed) && (tp.appliedRef <:< tp.selfType)
+        if validCtor(tp.cls) && isInstantiable then tp.cls
         else NoSymbol
       case tp: AppliedType =>
         samClass(tp.superType)

--- a/compiler/src/dotty/tools/dotc/core/Types.scala
+++ b/compiler/src/dotty/tools/dotc/core/Types.scala
@@ -5944,22 +5944,18 @@ object Types extends TypeUtils {
 
     def samClass(tp: Type)(using Context): Symbol = tp match
       case tp: ClassInfo =>
-        def zeroParams(tp: Type): Boolean = tp.stripPoly match
-          case mt: MethodType =>
-            val noArgsNeeded = mt.paramInfos match
-              case Nil => true
-              case info :: Nil => info.isRepeatedParam
-              case _ => false
-            noArgsNeeded && !mt.resultType.isInstanceOf[MethodType]
-          case et: ExprType => true
-          case _ => false
-        def validCtor(cls: Symbol): Boolean =
-          val ctor = cls.primaryConstructor
-          (!ctor.exists || zeroParams(ctor.info)) // `ContextFunctionN` does not have constructors
-          && (!cls.is(Trait) || validCtor(cls.info.parents.head.classSymbol))
+        val cls = tp.cls
+        def takesNoArgs(tp: Type) =
+          !tp.classSymbol.primaryConstructor.exists
+              // e.g. `ContextFunctionN` does not have constructors
+          || tp.applicableConstructors(Nil, adaptVarargs = true).lengthCompare(1) == 0
+              // we require a unique constructor so that SAM expansion is deterministic
+        val noArgsNeeded: Boolean =
+          takesNoArgs(tp)
+          && (!tp.cls.is(Trait) || takesNoArgs(tp.parents.head))
         def isInstantiable =
           !tp.cls.isOneOf(FinalOrSealed) && (tp.appliedRef <:< tp.selfType)
-        if validCtor(tp.cls) && isInstantiable then tp.cls
+        if noArgsNeeded && isInstantiable then tp.cls
         else NoSymbol
       case tp: AppliedType =>
         samClass(tp.superType)

--- a/compiler/src/dotty/tools/dotc/transform/ExpandSAMs.scala
+++ b/compiler/src/dotty/tools/dotc/transform/ExpandSAMs.scala
@@ -69,10 +69,12 @@ class ExpandSAMs extends MiniPhase:
           val tpe1 = collectAndStripRefinements(tpe)
           val Seq(samDenot) = tpe1.possibleSamMethods
           cpy.Block(tree)(stats,
-            AnonClass(List(tpe1),
-              List(samDenot.symbol.asTerm.name -> fn.symbol.asTerm),
-              refinements.toList
-            )
+            transformFollowingDeep:
+              AnonClass(List(tpe1),
+                List(samDenot.symbol.asTerm.name -> fn.symbol.asTerm),
+                refinements.toList,
+                adaptVarargs = true
+              )
           )
       }
     case _ =>

--- a/tests/neg/i15855.scala
+++ b/tests/neg/i15855.scala
@@ -1,0 +1,10 @@
+// crash.scala
+import scala.language.implicitConversions
+
+class MyFunction(args: String)
+
+trait MyFunction0[+R] extends MyFunction {
+  def apply(): R
+}
+
+def fromFunction0[R](f: Function0[R]): MyFunction0[R] = () => f() // error

--- a/tests/neg/i15855.scala
+++ b/tests/neg/i15855.scala
@@ -1,6 +1,3 @@
-// crash.scala
-import scala.language.implicitConversions
-
 class MyFunction(args: String)
 
 trait MyFunction0[+R] extends MyFunction {
@@ -8,3 +5,11 @@ trait MyFunction0[+R] extends MyFunction {
 }
 
 def fromFunction0[R](f: Function0[R]): MyFunction0[R] = () => f() // error
+
+class MyFunctionWithImplicit(implicit args: String)
+
+trait MyFunction0WithImplicit[+R] extends MyFunctionWithImplicit {
+  def apply(): R
+}
+
+def fromFunction1[R](f: Function0[R]): MyFunction0WithImplicit[R] = () => f() // error

--- a/tests/run/i15855.scala
+++ b/tests/run/i15855.scala
@@ -1,6 +1,3 @@
-// crash.scala
-import scala.language.implicitConversions
-
 class MyFunction(args: String*)
 
 trait MyFunction0[+R] extends MyFunction {

--- a/tests/run/i15855.scala
+++ b/tests/run/i15855.scala
@@ -1,0 +1,20 @@
+// crash.scala
+import scala.language.implicitConversions
+
+class MyFunction(args: String*)
+
+trait MyFunction0[+R] extends MyFunction {
+  def apply(): R
+}
+
+abstract class MyFunction1[R](args: R*):
+  def apply(): R
+
+def fromFunction0[R](f: Function0[R]): MyFunction0[R] = () => f()
+def fromFunction1[R](f: Function0[R]): MyFunction1[R] = () => f()
+
+@main def Test =
+  val m0: MyFunction0[Int] = fromFunction0(() => 1)
+  val m1: MyFunction1[Int] = fromFunction1(() => 2)
+  assert(m0() == 1)
+  assert(m1() == 2)

